Add governance document

sebmck opened the pull request:
This PR introduces a governance document that outlines the governance model for Rome. This includes the contributor model, code review, merging, and the consequences and process for code of conduct violations.

cc @romejs/contributors

VictorHom commented on docs/governance.md:
```diff
@@ -0,0 +1,25 @@
+
+- Someone who has proved themselves as an active contributor by being involved in project management or by submitting high quality pull requests.
+- There is no hard constraints. As the project evolves the bar for inclusion may be raised.
+- New contributors will be decided based on a general conseus by the existing core contributors.
```
```suggestion
- New contributors will be decided based on a general consensus by the existing core contributors.
```

sebmck commented:
An exception to the merge rule will be this PR where I'll wait for approval from all current contributors.

VictorHom commented on docs/governance.md:
```diff
@@ -0,0 +1,25 @@
+
+- Follow the [code of conduct](https://github.com/romejs/rome/blob/master/.github/CODE_OF_CONDUCT.md). Violations will result in your contributor status being revoked.
+- Don't abuse your repo access to push malicious code, or to edit or delete other peoples comments outside of moderation.
+- There are no expectations around activity. There may be a time where contributor privileges are restricted for inactivity however you will not lose org membership. Once you are added to the org you will be able t.
```
```suggestion
- There are no expectations around activity. There may be a time where contributor privileges are restricted for inactivity. However you will not lose org membership. Once you are added to the org you will be able t.
```

VictorHom commented on docs/governance.md:
```diff
@@ -0,0 +1,25 @@
+
+- Follow the [code of conduct](https://github.com/romejs/rome/blob/master/.github/CODE_OF_CONDUCT.md). Violations will result in your contributor status being revoked.
+- Don't abuse your repo access to push malicious code, or to edit or delete other peoples comments outside of moderation.
+- There are no expectations around activity. There may be a time where contributor privileges are restricted for inactivity however you will not lose org membership. Once you are added to the org you will be able t.
```
then I wasn't sure how you wanted to finish the sentence, so I can't offer a suggestion haha

sebmck replied:
🤦

sebmck commented:
Added two new sections:

## OpenCollective funds

- Usage of funds has yet to be decided.

## Governance changes

- Future changes to this document will require approval from all core contributors.

samccone commented on docs/governance.md:
```diff
@@ -0,0 +1,33 @@
+
+## What is a core contributor?
+
+- Someone who has proved themselves as an active contributor by being involved in project management or by submitting high quality pull requests.
```
Are there cons to adding anyone who lands a commit in rome to the owners file?

Pros:
* those who want to contribute can, and are not inhibited by process
* there is no gatekeeper, the project is owned by the community

Cons:
* The bad actor problem
 - mitigation: those acting against the goals of the project could be removed / banned from the project

* code-style / api design choices that go against the goals of the project
- mitigation: require no direct pushes to master, require all code to be code-reviewed by at least 1 additional person

sebmck replied:
The distinction between contributor and core contributor is necessary to minimize the amount of people who can authoritatively make decisions. How do you make any decisions when you need to get consensus amongst hundreds of people?

MylesBorins replied:
You can scale infinitely if you use [consensus-seeking](https://en.wikipedia.org/wiki/Consensus-seeking_decision-making)

Call for objections, minimal amount of review. This is what Node.js does. 1 - 2 signoff needed and 48 hours. Trust collaborators to do the right thing. Sometimes is doesn't work, but in 5 years that has happened very rarely.

FWIW I have very mixed feelings for "open open source"... e.g. give everyone the bit asap. Especially when we are talking about general supply chain security. I do think being a bit more liberal with permissions is good, but striking a balance is hard.

sebmck replied:
I've decided the current set of core contributors based on consistent high quality contributions. I would expect the bar to be raised after the project has a release and is more mature.

samccone commented on docs/governance.md:
```diff
@@ -0,0 +1,33 @@
+
+## Expectations
+
+- Follow the [code of conduct](https://github.com/romejs/rome/blob/master/.github/CODE_OF_CONDUCT.md). Violations will result in your contributor status being revoked.
```
> Violations will result in your contributor status being revoked.

can there be a public BAN_LIST in the project when individuals in violation are identified? This would act as a public record and remove any sort of backroom perception of the project

sebmck replied:
I'll add a moderation section that outlines this in more detail.

samccone commented on docs/governance.md:
```diff
@@ -0,0 +1,33 @@
+## Expectations
+
+- Follow the [code of conduct](https://github.com/romejs/rome/blob/master/.github/CODE_OF_CONDUCT.md). Violations will result in your contributor status being revoked.
+- Don't abuse your repo access to push malicious code, or to edit or delete other peoples comments outside of moderation.
```
one option here is to remove direct push from everyone except for org owners

sebmck replied:
Good point. I can just modify the branch protections. I'll update to clarify.

MylesBorins replied:
TBH these kinds of specific rules might be better in moderation guidelines rather than the governance. Governance documents can get large fast, especially as more process gets introduced.

Rules around how to remove someone ✅
All the reasons to remove them ❌

samccone commented on docs/governance.md:
```diff
@@ -0,0 +1,33 @@
+## Code review and merging
+
+- All code needs to go through pull requests and must pass status checks before being merged. If a PR is merged that breaks `master` due to conflicts then submit a new PR to fix it.
+- If a PR is against code that you have previously committed and is either small changes, bug fixes, or refactors, then you're free to merge it without any review. However if you don't feel confident in your changes then wait for approval from another core contributor.
```
> then you're free to merge it without any review. However if you don't feel ..

^ what is the downside of waiting for review? is it a bandwidth / latency problem of the core-team?

sebmck replied:
Yes. You should be able to drop a simple PR when you have time instead of making a big ordeal and waiting days or hours for an approval. It's about time optimization for everyone involved. I trust that everyone knows when it's appropriate to merge without approval.

MylesBorins commented on docs/governance.md:
```diff
@@ -0,0 +1,33 @@
+
+## Code review and merging
+
+- All code needs to go through pull requests and must pass status checks before being merged. If a PR is merged that breaks `master` due to conflicts then submit a new PR to fix it.
```
or revert?

Some of this stuff around "best practices" might be better fit in a CONTRIBUTING.md at some point. I know I know... I'm repeating myself...

When you can land code ✅
How to land it and what to do when it's broken ❌

sebmck replied:
We have some generated files and it's possible to have two PRs that are fine on their own but when merged one after another cause generated files to be in the wrong order. It's much easier to regenerate the files or fix it rather than having to revert then redo all of your changes as another PR.

MylesBorins replied:
also something that breaks master could be reverted rather than fixed 😇

MylesBorins commented on docs/governance.md:
```diff
@@ -0,0 +1,33 @@
+- All code needs to go through pull requests and must pass status checks before being merged. If a PR is merged that breaks `master` due to conflicts then submit a new PR to fix it.
+- If a PR is against code that you have previously committed and is either small changes, bug fixes, or refactors, then you're free to merge it without any review. However if you don't feel confident in your changes then wait for approval from another core contributor.
+- If you have shown yourself to be an owner of a particular area, whether it's by substantial contribution or prior discussion, you are free to merge it without any review despite PR size.
+- Make sure you have a consensus on changes. If you are adding a new feature then ensure that it has been discussed or approved on GitHub or Discord.
```
Ensuring consensus seems to go against the above rules not requiring sign off.

Obviously this will be a function of how many active collaborators you have and you don't want to create a bottleneck... but allowing people to carte blanche land PRs without review for areas they "own" doesn't seem inline with requiring consensus... unless you need to only build consensus between owners

sebmck commented:
I wasn't expecting for the governance guidelines to be so strict but I guess it makes sense. I'll make this document more detailed and list any changes in comments.

sebmck commented:
I was going to post a diff of my changes but I basically rewrote the whole document.

@MylesBorins @samccone Would appreciate your review here since I feel very out of my depth given the broad scope of the document now 😅 

@jamiebuilds @VictorHom Given the changes I'll disregard the previous approval.

sebmck commented:
I just wanted to acknowledge one flaw of the current governance. Given only one admin (me) there is trust placed on me to defer code of conduct violations involving mysdelf to the other core contributors.

There are some ways we can mitigate this:

 - Foundation ownership. I don't like this, I have a distrust for foundations, but it would mitigate it by offering another channel for escalation.
 - Add another admin who is a neutral third party.
 - Add another trusted admin involved with the project.

sebmck commented:
The governance model also omits guidelines around project direction decisions and decision consensus. I think we should leave it out for now and see how the project evolves because I feel it's too early for us to decide on that yet.

sebmck commented:
I have made the following changes:

 - Moved `CODE_OF_CONDUCT.md` from `.github` to the root
 - Moved `MODERATION.md` and `GOVERNANCE.md` to the root

I don't like polluting the top level files but this is important enough to warrant it.

cpojer commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,88 @@
+
+## Member Roles
+
+All members must follow the [code of conduct](./CODE_OF_CONDUCT.md). Consequences for member violations are detailed in [Moderation](#moderation).
```
"Code of Conduct"?

cpojer commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,88 @@
+
+### Core Contributor
+
+Core Contributors are those who have proved themselves as an active contributor by being involved in project management or by submitting high quality pull requests.
```
"proving themselves" sounds intimidating. Maybe "are those who have been an active contributor for a long period of time by being involved in project management or by submitting high quality pull requests."

You can also specify the period of time directly, like "one month" or "six months" or "two days".

sebmck replied:
It’s hard to quantify because there’s multiple variables: the contribution size, scope, project progress, frequency, and time. Maybe it’s worth being more vague since there needs to be a consensus amongst core contributors which essentially makes it a moving target anyway. Although even then it has the aura of “secret club”. It’s tricky.

cpojer replied:
I think specifying them as minimums is a good idea then.

cpojer commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,88 @@
+
+Core Contributors are those who have proved themselves as an active contributor by being involved in project management or by submitting high quality pull requests.
+
+There are no expectations around activity. There may be a time where contributor privileges are restricted or removed for inactivity. However you will not lose org membership.
```
Maybe clarify "once core contributor status is granted […]"

Also: "a time when"?

cpojer commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,88 @@
+
+There are no expectations around activity. There may be a time where contributor privileges are restricted or removed for inactivity. However you will not lose org membership.
+
+New contributors will be decided based on a general consensus by the existing core contributors.
```
Majority vote? Unanimous vote?

sebmck replied:
General consensus generally just means "hey is this good" "ok nobody has object for 2 days let's do it". I can clarify

cpojer commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,88 @@
+
+### Admin
+
+The amount of admins will be kept deliberately small. An admin has additional privileges over a core contributor including:
```
Why

cpojer commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,88 @@
+
+## Code review and merging
+
+- All code needs to go through pull requests and must pass status checks before being merged. If a PR is merged that breaks `master` due to the branch not being up to date, then it should either be reverted or a quick fix merged as a separate PR.
```
"through Pull Requests (PR)" – you are later using an abbreviation so gotta introduce it.

sebmck replied:
Fancy. I like it.

cpojer commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,88 @@
+## Code review and merging
+
+- All code needs to go through pull requests and must pass status checks before being merged. If a PR is merged that breaks `master` due to the branch not being up to date, then it should either be reverted or a quick fix merged as a separate PR.
+- If a PR is against code that you have previously committed and is either small changes, bug fixes, or refactors, then you're free to merge it without any review. However if you don't feel confident in your changes then wait for approval from another core contributor.
```
"then wait" -> "you can wait"

cpojer commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,88 @@
+
+- All code needs to go through pull requests and must pass status checks before being merged. If a PR is merged that breaks `master` due to the branch not being up to date, then it should either be reverted or a quick fix merged as a separate PR.
+- If a PR is against code that you have previously committed and is either small changes, bug fixes, or refactors, then you're free to merge it without any review. However if you don't feel confident in your changes then wait for approval from another core contributor.
+- If you are an owner of a particular area you are free to merge it without any review despite PR size.
```
wowza, that's a lot of power. Should this include some level of testing as a minimum requirement?

sebmck replied:
It can be suggested but I want to minimize hard requirements. Stigma and responsibility should balance it out. I'll expand on it. Thanks for the suggestion!

cpojer commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,88 @@
+- All code needs to go through pull requests and must pass status checks before being merged. If a PR is merged that breaks `master` due to the branch not being up to date, then it should either be reverted or a quick fix merged as a separate PR.
+- If a PR is against code that you have previously committed and is either small changes, bug fixes, or refactors, then you're free to merge it without any review. However if you don't feel confident in your changes then wait for approval from another core contributor.
+- If you are an owner of a particular area you are free to merge it without any review despite PR size.
+- If an area involves multiple owners then it's up to .
```
* "up to cpojer to act as a tie breaker" or whatever else you wanted to write here.

sebmck replied:
Meant to delete this since it's covered in the Owners section.

cpojer commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,88 @@
+
+## Moderation
+
+Users found to be in violation of the projects [code of conduct](./CODE_OF_CONDUCT.md) will be:
```
Is this a one-strike policy?

sebmck replied:
I don't know how to categorize code of conduct violations without minimizing violations. The wording here could definitely be better but I'm not sure how to allow discretion while properly conveying the gravity. Do you know how other projects do it?

cpojer commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,88 @@
+- Banned from the GitHub org and Discord server
+- Have their contributor status revoked (if applicable)
+- Have admin privileges revoked (if applicable)
+- Action listed in [`MODERATION.md`](./MODERATION.md). There may be some scenarios where discretion is required and some details omitted to protect individuals.
```
"protect individuals" or "protect and individual's privacy"?

cpojer commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,88 @@
+- Have admin privileges revoked (if applicable)
+- Action listed in [`MODERATION.md`](./MODERATION.md). There may be some scenarios where discretion is required and some details omitted to protect individuals.
+
+Code of conduct violations can be reported to <conduct@romejs.dev> which is listed in the [code of conduct](./CODE_OF_CONDUCT.md). This email address is monitored by admins.
```
The other doc made me think that it goes to "community leaders" that are separate from the admins. It may make sense to clarify that there. Also, bummer, because if I think an admin violated the CoC and this email goes out to all admins I don't really get protection as the person in question will also see my complaint and that it's coming from me. This may intimidate members from the community to report possible violations.

sebmck replied:
Yeah I raised that in [this comment](https://github.com/romejs/rome/pull/494#issuecomment-631871747). Other projects have this flaw but don't even acknowledge it nor have any thorough governance documentation.

Alternatively the Admins section can list emails of the admins to allow for disclosure to specific people. Also important to note that I don't see the project ever exceeding 3-4 admins.

sebmck replied:
I've made an update to include a list of admin contacts and clarification in the Code of Conduct intro.

sebmck commented:
Thanks for the review @cpojer! Much appreciated. I've addressed all but one of your comments. I feel like a one strike policy is probably necessary for CoC violations, which are typically egregious. It would be interesting to see the precedent among other projects.

sebmck commented:
Nevermind, all comments have been addressed. A one strike policy is appropriate given the gravity of Code of Conduct violations.

cpojer commented:
This is possibly the wrong place for this discussion but I think there is something to be said about reported violations that can be one-sided, like what if somebody's aim is to discredit somebody else after they got into a fight, or somebody sharing a really skewed perspective of what happened. I think it would be great to list the process in detail, including talking to the person who is accused of wrongdoing and identifying the root cause, instead of saying that a contributor will be banned after a violation is reported. While likely not the usual case (I do not want to discredit any future people who make use of the email address and report violations), it is possible that a report is wrong and the situation is explainable. Ideally the process will ensure that everyone is given privacy and space to deal with it instead of making quick decisions to resolve a tough situation.

(Note: this is me observing other projects, I have not been involved in such discussions or violations myself)

sebmck commented:
@cpojer The current document just says this:

> Moderation decisions and Code of Conduct violations will be discussed amongst the core contributors in private. Exceptions are if a possible violation involves a core contributor or admin:
>
> - Core contributor: Adjudicated amongst admins and other core contributors.
> - Admin: Adjudicated amongst the remaining admins and core contributors.
 
It does imply that there is some "undefined process" that does happen which definitely isn't ideal.

> reported violations that can be one-sided, like what if somebody's aim is to discredit somebody else after they got into a fight, or somebody sharing a really skewed perspective of what happened

I don't think this is actually a founded fear given that the described process already allows for discretion. There's no wording to imply that a report is equivalent to a violation.

If someone wants to have a go at defining a strict process I'm more than happy to add it.

sebmck commented:
After some more thinking I've limited moderation and CoC review to the admins. I can imagine it being extremely difficult to perform these kinds of sensitive reviews with dozens of people which is ultimately where I see the core contributors growing.

sebmck commented:
I've renamed Admins to Managers. Admins sounds really authoritarian while manager invokes a more managerial role which better conveys the intent.

TLadd commented on CODE_OF_CONDUCT.md:
```diff
@@ -1,5 +1,20 @@
+ - Sebastian McKenzie <sebmck@gmail.com> [GitHub](https://github.com/sebmck) [Twitter](https://twitter.com/sebmck)
+
+In the event that your report includes one of the people mentioned above please
+feel free to each out to one of the other people listed instead.
```
```suggestion
feel free to reach out to one of the other people listed instead.
```

TLadd commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,87 @@
+
+### Manager
+
+An manager has additional privileges over a core contributor. They serve to control and maintain sensitive project assets, and to act as tiebreakers in the event of disagreements. These additional privileges include:
```
```suggestion
A manager has additional privileges over a core contributor. They serve to control and maintain sensitive project assets, and to act as tiebreakers in the event of disagreements. These additional privileges include:
```

bitpshr commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,87 @@
+- If a PR is against code that you have previously committed and is either small changes, bug fixes, or refactors, then you're free to merge it without any review. However if you don't feel confident in your changes then you can wait for approval from another core contributor.
+- If you are an owner of a particular area you are free to merge it without any review despite PR size.
+- If after a PR is merged and there are comments or suggestions after the fact, allow yourself time to address them in a follow up PR. If you don't think you will be able to respond in a reasonable timeframe then create an issue to track.
+- If necessary, identity potential owners for PR review and approval.
```
```suggestion
- If necessary, identify potential owners for PR review and approval.
```

TLadd commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,87 @@
+
+This is a one-strike policy.
+
+Code of Conduct violations can be reported to <conduct@romejs.dev> which is listed in the [Code of Conduct](./CODE_OF_CONDUCT.md). This email address is monitored by managers. Alternatively email addresses for each manager is included in the Code of Conduct for private disclosements in the event of a report involving an manager.
```
```suggestion
Code of Conduct violations can be reported to <conduct@romejs.dev> which is listed in the [Code of Conduct](./CODE_OF_CONDUCT.md). This email address is monitored by managers. Alternatively email addresses for each manager is included in the Code of Conduct for private disclosements in the event of a report involving a manager.
```

TLadd commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,87 @@
+
+Code of Conduct violations can be reported to <conduct@romejs.dev> which is listed in the [Code of Conduct](./CODE_OF_CONDUCT.md). This email address is monitored by managers. Alternatively email addresses for each manager is included in the Code of Conduct for private disclosements in the event of a report involving an manager.
+
+Moderation decisions and Code of Conduct violations will be reviewed by the managers in private. If a violation involves an manager then it will be decided amongst the remaining managers. Review can be delegated to core contributors.
```
```suggestion
Moderation decisions and Code of Conduct violations will be reviewed by the managers in private. If a violation involves a manager then it will be decided amongst the remaining managers. Review can be delegated to core contributors.
```

ematipico commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,87 @@
+- If necessary, identity potential owners for PR review and approval.
+- Ensure that PR summary is detailed listing steps you took to verify, the rationale, and relevant issues and people involved in any prior discussion.
+- Ensure that PRs contain adequate tests and code comments for a future contributor to derive intent and modify the code safely.
+- You are welcome to use the `rome` repo for your WIP branches. However you should prefix branches with your username. ie. `git branch sebmck/feature`. Branches not involved in an active PR will be regularly pruned.
```
When a PR is opened from a fork, the username is prepended to the name of the branch `USERNAME:feature/my-awesome-feature`. Should this rule applied to these cases too? e.g. `ematipico:ematipico/feature/my-awesome-feature`

sebmck replied:
I don't think so. The only real reason for prepending the username is so we can easily tell from looking at the branch names who is responsible.

sebmck commented:
Thanks all for the review! Late night grammar isn't the best :tongue: Comments should have been addressed with the proposed suggestions.

rickhanlonii commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,87 @@
+
+### Core Contributor
+
+Core Contributors are those who have a history of consistent contributions, whether this pull requests, project management, or support.
```
```suggestion
Core Contributors include anyone with a history of consistent contributions, including but not limited to pull requests, project management, or support.
```

rickhanlonii replied:
Should this section have a list of privileges like the Managers section?

sebmck replied:
Good idea. I think we cover it in the "code review and merging" section but it would definitely be useful to reiterate in a summary form.

rickhanlonii commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,87 @@
+
+There are no expectations around activity. There may be a time where contributor privileges are restricted or removed for inactivity. Once core contributor status has been granted you will not lose org membership.
+
+New contributors will be decided based on a unanimous vote by the existing core contributors.
```
```suggestion
New core contributors will be added based on a unanimous vote by the existing core contributors.
```

rickhanlonii commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,87 @@
+- Ability to decide on moderation decisions involving core contributors
+- Access to the `*@romejs.dev` email address
+
+New managers will be decided based on a unanimous vote by the existing managers.
```
```suggestion
New managers will be added based on a unanimous vote by the existing managers.
```

rickhanlonii commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,87 @@
+
+### Manager
+
+A manager has additional privileges over a core contributor. They serve to control and maintain sensitive project assets, and to act as tiebreakers in the event of disagreements. These additional privileges include:
```
```suggestion
Managers have additional privileges over core contributors. Managers control and maintain sensitive project assets, and act as tiebreakers in the event of disagreements. These additional privileges include:
```

rickhanlonii commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,87 @@
+- [Paul Bouchon @bitpshr](https://github.com/bitpshr)
+- [Victor Hom @VictorHom](https://github.com/VictorHom)
+
+## Code review and merging
```
Consider splitting this into two different sections, one for guidelines to submitting PRs and one for guidelines to merging.

sebmck replied:
Good idea!

MylesBorins commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,87 @@
+
+There are no expectations around activity. There may be a time where contributor privileges are restricted or removed for inactivity. Once core contributor status has been granted you will not lose org membership.
+
+New core contributors will be added based on a unanimous vote by the existing core contributors.
```
If you are planning a unanimous vote here you might want to switch to consensus seeking... e.g. set time frame and no objections. Requiring a vote with all members having to vote can create serious friction if even a single person is not available. If the intent is "no objections" framing it that was might be more appropriate.

Also, would you be open to self nomination?

MylesBorins commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,87 @@
+- Ability to decide on moderation decisions involving core contributors
+- Access to the `*@romejs.dev` email address
+
+New managers will be added based on a unanimous vote by the existing managers.
```
Sam as above regarding unanimous vote vs. consensus seeking. But in this case, since there are likely very few managers, it may not be an issue. Although a manager being out on vacation or leave could in theory completely block this process based on current goernance.

MylesBorins commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,87 @@
+## Code review and merging
+
+- All code needs to go through Pull Requests (PR) and must pass status checks before being merged. If a PR is merged that breaks `master` due to the branch not being up to date, then it should either be reverted or a quick fix merged as a separate PR.
+- If a PR is against code that you have previously committed and is either small changes, bug fixes, or refactors, then you're free to merge it without any review. However if you don't feel confident in your changes then you can wait for approval from another core contributor.
```
I can see why you want to have this here but I very much am not a fan of not requiring review. Even a single reviewer help ensure that there is buyin and folks are not just pushing code that has 0 consensus around it

sebmck replied:
> However if you don't feel confident in your changes then you can wait for approval from another core contributor.

I don't think it's that bad. It places the responsibility on the core contributors. Publishing is performed by a manager so there isn't an opportunity for code to sneak into a release.

The only downside I can see is if things get reverted, and then the risk is placed on the core contributor who has potentially wasted their time, or needs to commit more. I think it will balance itself out.

We can always add more restrictions as the project matures.

MylesBorins commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,87 @@
+
+## OpenCollective fund allocation
+
+- Usage of funds has yet to be decided.
```
Do you have things you definitely don't want to use it for? Might be good spelling out anti-goals early

MylesBorins commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,87 @@
+
+## Governance changes
+
+- Future changes to this document will require approval from all core contributors.
```
Above you mention voting, here you state approval. I think it would be a good idea to pick a consensus model and document it. For changes to governance it is usually good to have a slightly larger timebox for changes.

sebmck commented on CODE_OF_CONDUCT.md:
```diff
@@ -1,5 +1,20 @@
+In the event that your report includes one of the people mentioned above please
+feel free to reach out to one of the other people listed instead.
+
+The moderation process for violations is listed in [`MODERATION.md`](./MODERATION.md).
```
I think it depends on the circumstance. If someone is deliberately vulgar and using their public identity then I do not feel it's inappropriate to "name and shame". Do you think discretion here is inappropriate?

MylesBorins replied:
I just don't think the public call out like that has any sort of positive outcome.

sebmck commented on CODE_OF_CONDUCT.md:
```diff
@@ -1,5 +1,20 @@
+
+ - Sebastian McKenzie <sebmck@gmail.com> [GitHub](https://github.com/sebmck) [Twitter](https://twitter.com/sebmck)
+
+In the event that your report includes one of the people mentioned above please
```
I don't think there is anyone who would currently fit the criteria given how early the project is.

sebmck commented:
I have addressed some of the comments above. Those that I've addressed have been marked as resolved. Unresolved comments are open for additional discussion.

**Changes:**

- Mention of fund usage for website and domain costs
- Merging and code review sections have been split up
- Process for approving governance changes and new core contributors has been clarified
- Responsibilities have been listed for core contributors

**Diff:**

```diff
--- a/GOVERNANCE.md
+++ b/GOVERNANCE.md
@@ -8,11 +8,21 @@ All members must follow the [Code of Conduct](CODE_OF_CONDUCT.md). Consequences

 ### Core Contributor

-Core Contributors include anyone with a history of consistent contributions, including but not limited to pull requests, project management, or support.
+Core Contributors include anyone with a history of consistent contributions, including but not limited to pull requests, project management, or support. These privileges include:
+
+ - Push access to the [Rome GitHub org](https://github.com/romejs), this includes all repos
+ - Contributor status on the [Rome Discord server](https://github.com/romejs)
+ - Ability to decide on new core contributors and governance changes

 There are no expectations around activity. There may be a time where contributor privileges are restricted or removed for inactivity. Once core contributor status has been granted you will not lose org membership.

-New core contributors will be added based on a unanimous vote by the existing core contributors.
+#### Induction
+
+Core contributors may either be nominated by another core contributor or can be self nominated. This can be done via message in the [`#contributors` Discord channel](https://discord.com/channels/678763474494423051/678763474930761739).
+
+New core contributors will be added based on a unanimous vote by the existing core contributors. Discussion and approval will take place in the [`#contributors-private` Discord channel](https://discord.com/channels/678763474494423051/712849311985041420).
+
+In the event of a rejection the nominated person will be privately given the requirements they have not met. Details of the discussion such as the names of those who objected will not be disclosed.

 ### Manager

@@ -24,10 +34,12 @@ Managers have additional privileges over core contributors. Managers control and
 - Publish access to the [`rome` npm package](https://www.npmjs.com/package/rome)
 - Domain registrar and DNS access to all `romejs.*` domains
 - Administration access to the `romejs.dev` Netlify account
-- Ability to decide on moderation decisions involving core contributors
+- Ability to decide on moderation decisions
 - Access to the `*@romejs.dev` email address

-New managers will be added based on a unanimous vote by the existing managers.
+#### Induction
+
+New managers will be added based on a unanimous vote by the existing managers. In the event that someone is unreachable then the decision will be deferred. Discussion and approval will be done in private.

 ### Owner

@@ -51,18 +63,21 @@ Members listed in alphabetical order.
 - [Paul Bouchon @bitpshr](https://github.com/bitpshr)
 - [Victor Hom @VictorHom](https://github.com/VictorHom)

-## Code review and merging
+## Code review

-- All code needs to go through Pull Requests (PR) and must pass status checks before being merged. If a PR is merged that breaks `master` due to the branch not being up to date, then it should either be reverted or a quick fix merged as a separate PR.
-- If a PR is against code that you have previously committed and is either small changes, bug fixes, or refactors, then you're free to merge it without any review. However if you don't feel confident in your changes then you can wait for approval from another core contributor.
 - If you are an owner of a particular area you are free to merge it without any review despite PR size.
 - If after a PR is merged and there are comments or suggestions after the fact, allow yourself time to address them in a follow up PR. If you don't think you will be able to respond in a reasonable timeframe then create an issue to track.
-- If necessary, identify potential owners for PR review and approval.
 - Ensure that PR summary is detailed listing steps you took to verify, the rationale, and relevant issues and people involved in any prior discussion.
 - Ensure that PRs contain adequate tests and code comments for a future contributor to derive intent and modify the code safely.
 - You are welcome to use the `rome` repo for your WIP branches. However you should prefix branches with your username. ie. `git branch sebmck/feature`. Branches not involved in an active PR will be regularly pruned.
 - If you are adding a new feature then ensure that it has been discussed or approved on GitHub or Discord.

+##  Merging
+
+- If necessary, identify potential owners for PR review and approval.
+- All code needs to go through Pull Requests (PR) and must pass status checks before being merged. If a PR is merged that breaks `master` due to the branch not being up to date, then it should either be reverted or a quick fix merged as a separate PR.
+- If a PR is against code that you have previously committed and is either small changes, bug fixes, or refactors, then you're free to merge it without any review. However if you don't feel confident in your changes then you can wait for approval from another core contributor.
+
 ## Moderation

 Users found to be in violation of the projects [Code of Conduct](./CODE_OF_CONDUCT.md) will be:
@@ -70,7 +85,7 @@ Users found to be in violation of the projects [Code of Conduct](./CODE_OF_CONDU
 - Banned from the GitHub org and Discord server
 - Have their contributor status revoked (if applicable)
 - Have manager privileges revoked (if applicable)
-- Action listed in [`MODERATION.md`](./MODERATION.md). There may be some scenarios where discretion is required and some details omitted to protect and individual's privacy.
+- Action listed in [`MODERATION.md`](./MODERATION.md). There may be some scenarios where discretion is required and some details omitted to protect an individual's privacy.

 This is a one-strike policy.

@@ -80,8 +95,9 @@ Moderation decisions and Code of Conduct violations will be reviewed by the mana

 ## OpenCollective fund allocation

-- Usage of funds has yet to be decided.
+- Funds will be allocated for project-specific services such as domain registration and website hosting.
+- Other usage of funds has yet to be decided.

 ## Governance changes

-- Future changes to this document will require approval from all core contributors.
+- Future changes to this document will require approval from code contributors. Time frame for acceptance will be unanimous approval or 7 days with no objections.
--
```

sebmck commented:
**Pending changes**

 - Public moderation disclosure
 - Code of Conduct violation review

sebmck commented:
Update: Haven't forgotten about this. I'll try and make some time in the next few days to complete the changes.

Kelbie commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,103 @@
+- [Eduardo Lopes @EduardoLopes](https://github.com/EduardoLopes)
+- [Florent Cailhol @ooflorent](https://github.com/ooflorent)
+- [Jamie Kyle @jamiebuilds](https://github.com/jamiebuilds)
+- [Kevin Kelbie](https://github.com/KevinKelbie)
```
I recently renamed my GitHub
```suggestion
- [Kevin Kelbie](https://github.com/Kelbie)
```

sebmck commented:
I've updated the governance document with the following:

- Manager has been renamed to director
- Voting rules have been defined
- Greater privileges for directors such as dictating high level project goals and the ability to veto a voting decision

I just need to add code of conduct review steps and this should be ready for finalization and final approval.

sebmck commented:
I have added the following moderation section. For posterity, please see the full document.

Notably the requirement for public disclosure has been removed.

------

## Moderation

Outlined below is the process for Code of Conduct violation reviews.

### Reporting

Anyone may report a violation. Violations can be reported in the following ways:

- In private, via <conduct@romejs.dev> which is listed in the [Code of Conduct](./CODE_OF_CONDUCT.md). This email address is monitored by all directors.
- In private, via email to one or more dcirectors.
- In private, via direct message to a project director on Discord
- In public, via a GitHub comment (mentioning `@romejs/directors`).
- In public, via the project Discord server.

### Who gets involved?

Each report will be assigned reviewers. These will initially be all project [directors](#directors).

In the event of any conflict of interest - ie. directors who are personally connected to a situation, they must immediately recuse themselves.

At request of the reporter and if deemed appropriate by the reviwers, another netural third-party may be involved in the review and decision process.

### Review

If a report doesn’t contain enough information, the reviewers will strive to obtain all relevant data before acting.

The reviewers will then review the incident and determine, to the best of their ability:

- What happened.
- Whether this event constitutes a Code of Conduct violation.
- Who, if anyone, was involved in the violation.
- Whether this is an ongoing situation.

The reviewers should aim to have a resolution agreed very rapidly; if not agreed within a week, they will inform the parties of the planned date.

### Resolution

Responses will be determined by the reviewers on the basis of the information gathered and of the potential consequences. It may include:

- taking no further action
- issuing a reprimand (private or public)
- asking for an apology (private or public)
- permanent ban from the GitHub org and Discord server
- revoked contributor status

sebmck commented:
This document is now comprehensive and ready for final approval and comments.

Future approvals will be not be re-requested. It will be the responsibility of those approving now to monitor this PR for changes and withdraw their approval if necessary.

I will merge this once there is unanimous approval from all current core contributors. Although per the rules currently outlined in this document, future governance changes will only require a simple majority (over 50%).

yassere commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,154 @@
+Anyone may report a violation. Violations can be reported in the following ways:
+
+- In private, via <conduct@romejs.dev> which is listed in the [Code of Conduct](./CODE_OF_CONDUCT.md). This email address is monitored by all directors.
+- In private, via email to one or more dcirectors.
```
Small typo: dcirectors -> directors

yassere commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,154 @@
+
+In the event of any conflict of interest - ie. directors who are personally connected to a situation, they must immediately recuse themselves.
+
+At request of the reporter and if deemed appropriate by the reviwers, another netural third-party may be involved in the review and decision process.
```
netural -> neutral

MylesBorins commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,154 @@
+
+In the event of a rejection the nominated person will be privately given the requirements they have not met. Details of the discussion such as the names of those who objected will not be disclosed.
+
+### Director
```
Is there a reason to use the term "Director" rather than the classical "Technical Steering Committee"?

Usually in a foundation setting "Director" has a specific meaning which includes accepting certain legal and financial risk associated with fiduciary duty. This feedback is a bit pedantic and can be ignored... but definitely interested in the though process here.

sebmck replied:
I don't expect there to be more than three "directors" so calling it a committee I think carries more weight. I am not super opinionated on the terminology, previously I used "manager" but I think that carries the wrong connotations and can be demeaning.

In practice I expect the distinction between director and contributor to only be relevant because one has publishing rights (something we can also change in the future).

I would consider the additional "powers" that a director has such as setting high-level project goals and veto ability to be really rare and only a last resort. I really don't anticipate it having to be used at all but having it gives the project more accountability.

MylesBorins replied:
What about "Steward"

> (of an official) supervise arrangements or keep order at (a large public event).

sebmck replied:
I like it!

bitpshr commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,154 @@
+
+In the event of any conflict of interest - ie. directors who are personally connected to a situation, they must immediately recuse themselves.
+
+At request of the reporter and if deemed appropriate by the reviwers, another neutral third-party may be involved in the review and decision process.
```
```suggestion
At request of the reporter and if deemed appropriate by the reviewers, another neutral third-party may be involved in the review and decision process.
```

sebmck commented:
Renamed director to steward per @MylesBorins' recommendation. Thank you!

VictorHom commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,154 @@
+- [Olivier Dusabimana @diokey](https://github.com/diokey)
+- [Paul Bouchon @bitpshr](https://github.com/bitpshr)
+- [Victor Hom @VictorHom](https://github.com/VictorHom)
+
```
Adding on @yassere to the list when you have the time @sebmck

sebmck commented:
cc @ooflorent waiting on your review!

sebmck commented:
🙌
